### ISSUE TITLE
[Crash] Intermittent SIGABRT in TabTrayViewController.handleNotifications when closing tabs

### ISSUE THREAD

**Issue by cpeterso** [mozilla-mobile/firefox-ios#10447]
## Steps to reproduce

After updating to the recent Nightly version v99 (8544), Firefox crashes about once or twice every day. The STR is something related to closing a tab or switching from Firefox to another app. I think these crashes are a new regression in v99. I didn't have any crashes in v98 (8233).

I have some Client-2022-04-XX-XXXXXX.ips crash logs from my iPhone I can share. I'd rather not post them in a public GitHub issue, in case there is any private user information in them.

### Expected behavior

No crash.

### Actual behavior

The crash looks like heap corruption or a double-free. Here's the crash stack trace from a Client ips file:

```
Coalition:           org.mozilla.ios.FirefoxBeta [773]

Date/Time:           2022-04-09 18:20:56.2951 -0700
Launch Time:         2022-04-09 16:15:52.7311 -0700
OS Version:          iPhone OS 15.4.1 (19E258)
Release Type:        User
Baseband Version:    2.53.01
Report Version:      104

Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Triggered by Thread:  14

Application Specific Information:
abort() called

Thread 0 name:   Dispatch queue: com.apple.main-thread
Thread 0:
0   libsystem_kernel.dylib        	       0x21e5f9b78 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x2582683bc pthread_kill + 268
2   libsystem_c.dylib             	       0x1f1c1451c abort + 168
3   libsystem_malloc.dylib        	       0x1f8d50a04 malloc_vreport + 552
4   libsystem_malloc.dylib        	       0x1f8d50c9c malloc_zone_error + 100
5   libsystem_malloc.dylib        	       0x1f8d4b0dc nanov2_allocate_from_block + 568
6   libsystem_malloc.dylib        	       0x1f8d4a164 nanov2_allocate + 128
7   libsystem_malloc.dylib        	       0x1f8d4a750 nanov2_calloc + 160
8   libsystem_malloc.dylib        	       0x1f8d38e58 _malloc_zone_calloc + 88
9   libobjc.A.dylib               	       0x1ff7853a0 _objc_rootAllocWithZone + 40
10  UIKitCore                     	       0x1e9193bd0 -[UITraitCollection imageConfiguration] + 32
11  UIKitCore                     	       0x1e90b0c44 +[UIImageConfiguration _completeConfiguration:fromConfiguration:] + 444
12  UIKitCore                     	       0x1e923d40c +[UIImage imageNamed:inBundle:withConfiguration:] + 180
13  Client                        	       0x100b2b498 TabTrayViewController.handleNotifications(_:) + 788
14  Client                        	       0x100b2b61c @objc TabTrayViewController.handleNotifications(_:) + 128
15  CoreFoundation                	       0x1e6ac3324 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 28
16  CoreFoundation                	       0x1e6b5fac4 ___CFXRegistrationPost_block_invoke + 52
17  CoreFoundation                	       0x1e6b32cc0 _CFXRegistrationPost + 456
18  CoreFoundation                	       0x1e6ad939c _CFXNotificationPost + 728
19  Foundation                    	       0x1e82a0704 -[NSNotificationCenter postNotificationName:object:userInfo:] + 96
20  Client                        	       0x100b7f2f4 protocol witness for TabManagerDelegate.tabManager(_:didRemoveTab:isRestoring:) in conformance GridTabViewController + 76
21  Client                        	       0x100dc85d0 TabManager.removeTab(_:flushToDisk:) + 1220
22  Client                        	       0x100dc7ab0 TabManager.removeTab(_:) + 324
23  Client                        	       0x100bfd230 specialized closure #1 in TabDisplayManager.closeActionPerformed(forCell:) + 484
24  Client                        	       0x100bfd750 specialized TabDisplayManager.setupSearchTermGroupsAndFilteredTabs(tabsToBuildFrom:completion:) + 1196
25  Client                        	       0x100bfe318 specialized TabDisplayManager.getTabsAndUpdateInactiveState(completion:) + 636
26  Client                        	       0x100bf3678 TabDisplayManager.closeActionPerformed(forCell:) + 584
27  Client                        	       0x100b85174 specialized GridTabViewController.swipeAnimator(_:viewWillExitContainerBounds:) + 664
28  Client                        	       0x100dad7b8 SwipeAnimator.animateAwayWithVelocity(_:speed:) + 244
29  Client                        	       0x100dadfa4 SwipeAnimator.didPan(_:) + 864
30  Client                        	       0x100dadff0 @objc SwipeAnimator.didPan(_:) + 52
31  UIKitCore                     	       0x1e90e64e8 -[UIGestureRecognizerTarget _sendActionWithGestureRecognizer:] + 56
32  UIKitCore                     	       0x1e90afb58 _UIGestureRecognizerSendTargetActions + 116
33  UIKitCore                     	       0x1e907903c _UIGestureRecognizerSendActions + 288
34  UIKitCore                     	       0x1e90b1ea0 -[UIGestureRecognizer _updateGestureForActiveEvents] + 612
35  UIKitCore                     	       0x1e906aa58 _UIGestureEnvironmentUpdate + 2008
36  UIKitCore                     	       0x1e909de24 -[UIGestureEnvironment _updateForEvent:window:] + 772
37  UIKitCore                     	       0x1e90ab0e4 -[UIWindow sendEvent:] + 4380
38  UIKitCore                     	       0x1e9258ae4 -[UIApplication sendEvent:] + 824
39  UIKitCore                     	       0x1e907e9ec __dispatchPreprocessedEventFromEventQueue + 7896
40  UIKitCore                     	       0x1e907371c __processEventQueue + 6616
41  UIKitCore                     	       0x1e9ec3470 updateCycleEntry + 176
42  UIKitCore                     	       0x1e96e1084 _UIUpdateSequenceRun + 84
43  UIKitCore                     	       0x1e9d66cb0 schedulerStepScheduledMainSection + 144
44  UIKitCore                     	       0x1e9d66478 runloopSourceCallback + 92
45  CoreFoundation                	       0x1e6b53f04 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28
46  CoreFoundation                	       0x1e6b64c90 __CFRunLoopDoSource0 + 208
47  CoreFoundation                	       0x1e6a9e184 __CFRunLoopDoSources0 + 268
48  CoreFoundation                	       0x1e6aa3b4c __CFRunLoopRun + 828
49  CoreFoundation                	       0x1e6ab76b8 CFRunLoopRunSpecific + 600
50  GraphicsServices              	       0x202b51374 GSEventRunModal + 164
51  UIKitCore                     	       0x1e941ce88 -[UIApplication _run] + 1100
52  UIKitCore                     	       0x1e919e5ec UIApplicationMain + 364
53  Client                        	       0x100aa2938 main + 260
54  dyld                          	       0x101ec9ce4 start + 520
```

### Device & build information

* Device: iPhone 12 running iOS 15.4.1
* Build version: Nightly v99 (8544)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-4080)

**Comment by cpeterso** [comment]
Firefox (now version 103.0 (13910)) still crashes for me about 1-2 times a day when closing a tab or switching to another app. I don't know if I'm seeing the same crash. My client .ips crash logs don't mention `-[UITraitCollection imageConfiguration]` or `TabTrayViewController.handleNotifications`.

Here's an excerpt:

CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION

**Comment by dnarcese** [comment]
@cpeterso are you still seeing this crash in recent versions?

**Comment by cpeterso** [comment]
> @cpeterso are you still seeing this crash in recent versions?

Yes. I saw this crash once yesterday (v104), three times on Tuesday (v103), and twice over the weekend. I have client .ips crash logs I can share.

Roux suggested I file a new issue because the crash stack trace in my client .ips crash logs looked different from this bug, though the STR is the same. So I filed https://github.com/mozilla-mobile/firefox-ios/issues/11485
